gtfs-validator/main/src/test/java/org/mobilitydata/gtfsvalidator/validator/RouteUniqueNamesValidatorTest.java

```diff
@@ -0,0 +1,206 @@
+  public void routeOnlyLongNameEqualsAnotherRouteBothLongShortNames() {
+    final NoticeContainer noticeContainer = new NoticeContainer();
+    RouteUniqueNamesValidator validator = new RouteUniqueNamesValidator();
+
+    List<GtfsRoute> routes = new ArrayList<>();
+    routes.add(routeOnlyLongName_abcd);
+    routes.add(routeLongName_ab_ShortName_cd);
+    validator.routeTable = GtfsRouteTableContainer.forEntities(routes, noticeContainer);
+
+    validator.validate(noticeContainer);
+    assertThat(noticeContainer.getNotices())
+        .containsExactly(
+            new DuplicateRouteLongNameShortNameCombinationNotice(
+                "routeOnlyLongName_abcd", "abcd", "", "routeLongName_ab_ShortName_cd", "ab", "cd"));
+  }
+
+  @Test
+  public void routeOnlyShortNameEqualsAnotherRouteBothLongShortNames() {
+    final NoticeContainer noticeContainer = new NoticeContainer();
+    RouteUniqueNamesValidator validator = new RouteUniqueNamesValidator();
+
+    List<GtfsRoute> routes = new ArrayList<>();
+    routes.add(routeOnlyShortName_abcd);
+    routes.add(routeLongName_ab_ShortName_cd);
+    validator.routeTable = GtfsRouteTableContainer.forEntities(routes, noticeContainer);
+
+    validator.validate(noticeContainer);
+    assertThat(noticeContainer.getNotices())
+        .containsExactly(
+            new DuplicateRouteLongNameShortNameCombinationNotice(
+                "routeOnlyShortName_abcd",
+                "",
+                "abcd",
+                "routeLongName_ab_ShortName_cd",
+                "ab",
+                "cd"));
+  }
+
+  @Test
+  public void routeOnlyShortNameDifferentFromAnotherRouteBothLongShortNames() {
+    final NoticeContainer noticeContainer = new NoticeContainer();
+    RouteUniqueNamesValidator validator = new RouteUniqueNamesValidator();
+
+    List<GtfsRoute> routes = new ArrayList<>();
+    routes.add(routeOnlyShortName_abcd);
+    routes.add(routeLongName_cd_ShortName_ab);
+    validator.routeTable = GtfsRouteTableContainer.forEntities(routes, noticeContainer);
+
+    validator.validate(noticeContainer);
+    assertThat(noticeContainer.getNotices()).isEmpty();
+  }
+
+  @Test
+  public void routeBothLongShortNamesEqualsAnotherRouteBothLongShortNames() {
+    final NoticeContainer noticeContainer = new NoticeContainer();
+    RouteUniqueNamesValidator validator = new RouteUniqueNamesValidator();
+
+    List<GtfsRoute> routes = new ArrayList<>();
+    routes.add(routeLongName_ab_ShortName_cd);
+    routes.add(routeLongName_abc_ShortName_d);
+    validator.routeTable = GtfsRouteTableContainer.forEntities(routes, noticeContainer);
+
+    validator.validate(noticeContainer);
+    assertThat(noticeContainer.getNotices())
+        .containsExactly(
+            new DuplicateRouteLongNameShortNameCombinationNotice(
+                "routeLongName_ab_ShortName_cd",
+                "ab",
+                "cd",
+                "routeLongName_abc_ShortName_d",
+                "abc",
+                "d"));
```

Comment:
Is this a desired behaviour?

Reply:
Yes. Because a route with LongName "ab" and ShortName "cd" has the name combination "abcd", and it is the same with a route with LongName "abc" and ShortName "d", which should generate a warning for DuplicateRouteLongNameShortNameCombination.

Reply:
I am not sure why this requirement needs to be true.
Looking at https://developers.google.com/transit/gtfs/reference#routestxt There does not seem to be such explicit requirement.
If the old validator implemented such a rule we can enforce it for compatibility.

@tripd22 Can you please check if we want to enforce such requirement?

Reply:
Yes, the rule logic comes from "UniqueRouteLongNameRouteShortNameCombination" Warning in the old validator. 
I thought it is meaningful because we only check the uniqueness of short name and long name of route respectively for the same agency _(in the left two rules that I have not implemented in the validator)_, while never checking what will happen if the agency is not the same. 
This rule makes sure the combination of information from route's name is unique and meaningful. Also, when the route only has long name or short name (which will happen from time to time), the only name is unique among all names (both long and short ones).

Reply:
I think what we actually want to do is ensure that the combination of short name + long name + route type + agency id is unique.

This is what our C++ validator currently checks (though it used to check just the short name + long name combination). However, it doesn't concatenate the strings to do this. Instead, it basically hashes the 4 individual pieces of data together, and ensures the hash is unique.

So {"ab", "cd", 1, "agency_a"} and {"abc", "d", 1, "agency_a"} would be considered different.

You can probably use a custom class that holds these pieces of data to replicate this, and hash that (unless you know a better approach @jupi3200? I'll send you the C++ code snippet where we avoid creating a data structure.).


Note: The old mobility data validator also tests that short name is unique _and_ long name is unique. Our C++ validator doesn't test for this, and it isn't part of the spec, so I don't think we need to do this.

Reply:
Sure. I will then wait to adjust the validator logic :) rather than further implementing the other two rules.

Reply:
@jupi3200 suggested [autovalue](https://www.baeldung.com/introduction-to-autovalue), which looks like a better approach than mine!

Reply:
`Objects.hash(long_name, short_name, route_type, agency_id)` might be useful as well - that way you don't need to create a new object to get a hash of the values.

Reply:
Thank you! I'm going to change to that.

Reply:
I used this idea and changed the validator. Because almost all files were changed, I created another new PR for the new version - PR #28 .